Add example for group-by and collect (#6923)
ballerina-platform/ballerina-dev-website: swan-lake/featured-use-cases/work-with-data-using-queries-in-ballerina.md

Countries with number of Recovered patients:[["Afghanistan",146084],["USA",43892277],["India",33892279]]
Found erroneous entries for countries: ["Sri Lanka","India"]
```
## Group COVID-19 data by continent

> ### Create the groupDeathsByContinent function

Reviewer: ```suggestion
### Create the `groupDeathsByContinent` function
```

> To difine a function to group the data by continent and get the total number of deaths by continent, add the code below to the `main.bal` file.

Reviewer 2: ```suggestion
To define a function that groups the data by continent and retrieves the total number of deaths per continent, add the following code to the main.bal file.
```

```ballerina
public function groupDeathsByContinent(table<CovidEntry> dataTable) returns record {|string continent; decimal deaths;|}[] {
    return from var {continent, deaths} in dataTable
        group by continent
        select {continent, deaths: sum(deaths)};
}
```

In this code,

> 1. You use `group by` clause to group the input.

Reviewer: ```suggestion
1. You use the `group by` clause to group the input.
```
These should be unordered bullet points as they aren't sequential.

> 2. Since the `grouping-key` is `continent` the input table is grouped by `continent` in the function `groupDeathsByContinent`.

Reviewer 3: `Have we introduced what `grouping-key` is? If not, users might be confused no knowing what it is.

> - You use the `group by` clause to group the input. After the `group by` you can find the `grouping-key`, in this example `continent` is the `grouping-key`. The `grouping-key` is used to group the input and each group, the `grouping-key` is unique.

Reviewer 2: `grouping-key` is unique or `grouping-key` should be unique.
what is the best one?

Author: I think `is unique` is better.

3. After creating groups based on `continent`, `deaths` becomes a sequence value for each `continent`.

> 4. In order to get the total number of deaths per `continent`, you can call the `sum` function.

Reviewer: ```suggestion
4. You can call the `sum` function to get the total number of deaths per `continent`.
```

    {iso_code: "IND", country: "India", continent: "Asia", cases: 80808350, deaths: 980976, recovered: 33892279, active: 35035095}
];

public function filterCountriesByCases(table<CovidEntry> dataTable, decimal noOfCases) returns string[] {
    string[] filteredCountries = from CovidEntry entry in dataTable
        where entry.cases > noOfCases
        select entry.country;
    return filteredCountries;
}

public function findCountriesByHighestNoOfDeaths(table<CovidEntry> dataTable, int n) returns [string, decimal][] {
    [string, decimal][] countriesWithDeaths = from CovidEntry entry in dataTable
        order by entry.deaths descending
        limit n
        select [entry.country, entry.deaths];
    return countriesWithDeaths;
}

public function findRecoveredPatientsOfCountries(table<CovidEntry> dataTable, string[] countries) returns [string, decimal][] {
    [string, decimal][] countriesWithRecovered = from CovidEntry entry in dataTable
        join string country in countries on entry.country equals country
        select [entry.country, entry.recovered];
    return countriesWithRecovered;
}

public function printErroneousData(table<CovidEntry> dataTable) {

> 

Reviewer 2: ```suggestion
```

    string[] countries = from CovidEntry entry in dataTable
        let decimal sum = entry.recovered + entry.deaths + entry.active
        where entry.cases != sum
        select entry.country;

    if countries.length() > 0 {
        io:println("Found erroneous entries for countries: ", countries);
    }
}

public function groupDeathsByContinent(table<CovidEntry> dataTable) returns record {|string continent; decimal deaths;|}[] {
    return from var {continent, deaths} in dataTable
        group by continent
        select {continent, deaths: sum(deaths)};
}

public function main() {
    string[] countries = filterCountriesByCases(covidTable, 10000000);
    io:println("Countries with more than 10 million cases: ", countries);

    [string, decimal][] countriesWithDeaths = findCountriesByHighestNoOfDeaths(covidTable, 3);

>     io:println("Countries with highest deaths:", countriesWithDeaths);

Reviewer 2: ```suggestion
    io:println("Countries with highest deaths: ", countriesWithDeaths);
```

    string[] c = ["USA", "India", "Afghanistan"];
    [string, decimal][] countriesWithRecovered = findRecoveredPatientsOfCountries(covidTable, c);

>     io:println("Countries with number of Recovered patients:", countriesWithRecovered);

Reviewer 2: ```suggestion
    io:println("Countries with number of Recovered patients: ", countriesWithRecovered);
```

    printErroneousData(covidTable);

    record {|string continent; decimal deaths;|}[] deathsByContinent = groupDeathsByContinent(covidTable);
    io:println("Total number of deaths by continent: ", deathsByContinent);
}
```

### Run the package for grouping

In the terminal, navigate to the `query_expressions` directory, and execute the command below to run the service package.

```bash
$ bal run
```

You view the output below.

```bash
Compiling source
	    .../query_expressions:0.1.0

Running executable

Compiling source
	kavindu/query_expressions:0.1.0

Running executable

Countries with more than 10 million cases: ["USA","India"]
Countries with highest deaths:[["India",980976],["USA",880976],["Afghanistan",7386]]
Countries with number of Recovered patients:[["Afghanistan",146084],["USA",43892277],["India",33892279]]
Found erroneous entries for countries: ["Sri Lanka","India"]
Total number of deaths by continent: [{"continent":"Asia","deaths":993605},{"continent":"North America","deaths":880976}]
```

> ## Get total COVID-19 deaths

Reviewer: ```suggestion
## Get the total COVID-19 deaths
```

> ### Create the getTotalNumberOfDeaths function

Reviewer: ```suggestion
### Create the `getTotalNumberOfDeaths` function
```

To define a function to get total number of deaths, add the code below to the `main.bal` file.

```ballerina
public function getTotalNumberOfDeaths(table<CovidEntry> dataTable) returns decimal {
    return from var {deaths} in dataTable
            collect sum(deaths);
}
```

In this code,
1. You use `collect` clause to collect all input frames into one group.
2. Since `deaths` becomes a sequence value, you can call the `sum` function in order to get the total number of deaths.

> ### Update the main function for collect

Reviewer: ```suggestion
### Update the main function for collecting
```

To call the `getTotalNumberOfDeaths` function within the main function to find the total number of deaths, add the code below to the main function of the `main.bal` file.

> ### The complete code with the collect function

Reviewer: ```suggestion
### The complete code with the collecting function
```

        group by continent
        select {continent, deaths: sum(deaths)};
}

public function getTotalNumberOfDeaths(table<CovidEntry> dataTable) returns decimal {
    return from var {deaths} in dataTable
            collect sum(deaths);
}

public function main() {
    string[] countries = filterCountriesByCases(covidTable, 10000000);
    io:println("Countries with more than 10 million cases: ", countries);

    [string, decimal][] countriesWithDeaths = findCountriesByHighestNoOfDeaths(covidTable, 3);
    io:println("Countries with highest deaths:", countriesWithDeaths);

    string[] c = ["USA", "India", "Afghanistan"];
    [string, decimal][] countriesWithRecovered = findRecoveredPatientsOfCountries(covidTable, c);
    io:println("Countries with number of Recovered patients:", countriesWithRecovered);

    printErroneousData(covidTable);

    record {|string continent; decimal deaths;|}[] deathsByContinent = groupDeathsByContinent(covidTable);
    io:println("Total number of deaths by continent: ", deathsByContinent);

    decimal totalDeaths = getTotalNumberOfDeaths(covidTable);
    io:println("Total number of deaths: ", totalDeaths);
}
```

> ### Run the package for collect

Reviewer: ```suggestion
### Run the package for collecting
```
